instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/ParameterizedClass.java

```diff
@@ -0,0 +1,197 @@
+    for (int i = 0; i < interfaceTypes.length; i++) {
+      parameterizedClasses[i] = withMappedTypeArguments(interfaceTypes[i]);
+    }
+    return parameterizedClasses;
+  }
+
+  public Optional<ParameterizedClass> findParameterizedSuperclass(Class<?> superClass) {
+    Objects.requireNonNull(superClass);
+    return findParameterizedSuperclassImpl(this, superClass);
+  }
+
+  private static Optional<ParameterizedClass> findParameterizedSuperclassImpl(
+      ParameterizedClass current, Class<?> superClass) {
+
+    if (current == null) {
+      return Optional.empty();
+    }
+    if (current.rawClass.equals(superClass)) {
+      return Optional.of(current);
+    }
+    if (superClass.isInterface()) {
+      for (ParameterizedClass interfaceType : current.getParameterizedInterfaces()) {
+        if (interfaceType.rawClass.equals(superClass)) {
+          return Optional.of(interfaceType);
+        }
+      }
+    }
+    return findParameterizedSuperclassImpl(current.getParameterizedSuperclass(), superClass);
+  }
+
+  private ParameterizedClass withMappedTypeArguments(Type superType) {
+    if (superType instanceof ParameterizedType) {
+      ParameterizedType parameterizedSuperType = (ParameterizedType) superType;
+      TypeVariable<?>[] typeParameters = rawClass.getTypeParameters();
+      return withMappedTypeArguments(parameterizedSuperType, typeParameters, typeArguments);
+    } else if (superType != null) {
+      return ParameterizedClass.of(superType);
+    }
+    return null;
+  }
+
+  private static ParameterizedClass withMappedTypeArguments(
+      ParameterizedType parameterizedSuperType,
+      TypeVariable<?>[] typeParameters,
+      Type[] typeArguments) {
+    Type[] superTypeArguments = parameterizedSuperType.getActualTypeArguments();
+
+    for (int i = 0; i < superTypeArguments.length; i++) {
+      Type superTypeArgument = superTypeArguments[i];
+      if (superTypeArgument instanceof TypeVariable) {
+        TypeVariable<?> superTypeVariable = (TypeVariable<?>) superTypeArgument;
+        superTypeArguments[i] =
+            mapTypeVariableToArgument(superTypeVariable, typeParameters, typeArguments);
+      }
+    }
+    return new ParameterizedClass(parameterizedSuperType, superTypeArguments);
+  }
+
+  private static Type mapTypeVariableToArgument(
+      TypeVariable<?> superTypeVariable, TypeVariable<?>[] typeParameters, Type[] typeArguments) {
+    for (int i = 0; i < typeParameters.length; i++) {
+      if (equalsTypeVariable(superTypeVariable, typeParameters[i])) {
+        return typeArguments[i];
+      }
+    }
+    return superTypeVariable;
+  }
+
+  private static boolean equalsTypeVariable(TypeVariable<?> left, TypeVariable<?> right) {
+    if (left == right) {
+      return true;
+    } else if (left == null || right == null) {
+      return false;
+    }
+    return left.getGenericDeclaration().equals(right.getGenericDeclaration())
+        && left.getName().equals(right.getName());
```

Comment:
Can do.  This was ported from Guava TypeToken API and this is the comment they use on the `TypeVariableKey` wrapper type they use to compare `TypeVariable<?>` identity:

```java
  /**
   * Wraps around {@code TypeVariable<?>} to ensure that any two type variables are equal as long as
   * they are declared by the same {@link java.lang.reflect.GenericDeclaration} and have the same
   * name, even if their bounds differ.
   *
   * <p>While resolving a type variable from a {@code var -> type} map, we don't care whether the
   * type variable's bound has been partially resolved. As long as the type variable "identity"
   * matches.
   *
   * <p>On the other hand, if for example we are resolving {@code List<A extends B>} to {@code
   * List<A extends String>}, we need to compare that {@code <A extends B>} is unequal to {@code <A
   * extends String>} in order to decide to use the transformed type instead of the original type.
   */
```